### PR TITLE
Add BSHD tests for context parallel

### PR DESCRIPTION
**Pull request by pstjohn**
Since padded-THD context parallelism is only supported on datacenter hardware, we should add BSHD CP examples to be able to test this code on workstations and in CI

This branch includes #1400 

Closes BIO-13

### CODE REVIEW

File: bionemo-recipes/recipes/llama3_native_te/checkpoint.py [hunk 1]
```diff
@@ -328,13 +328,15 @@ def save_checkpoint_fsdp2(
...
         logger.info(f"Saved FSDP2 dataloader to {ckpt_path}")
 
-    # If we're using asynchronous checkpointing, make sure we only have one checkpoint future at a time.
-    if async_save and "fsdp2" in _ckpt_futures and _ckpt_futures["fsdp2"] is not None:
-        _ckpt_futures["fsdp2"].result()
-
     state_dict = {"app": AppState(model=model, optimizer=optimizer, scheduler=scheduler, step=step, epoch=epoch)}
-    ckpt_save_func = dcp_async_save if async_save else dcp_save
-    _ckpt_futures["fsdp2"] = ckpt_save_func(state_dict, checkpoint_id=checkpoint_path, process_group=process_group)
+    if async_save:
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama3_native_te/checkpoint.py:332]: This kinda doesn't fit with this MR. Wondering what's going on here

File: bionemo-recipes/recipes/llama3_native_te/train_fsdp2_cp.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+    perf_logger = PerfLogger(dist_config, args)
+
+    gc.collect()
+    torch.cuda.empty_cache()
+
+    # Training loop
+    logger.info(f"Starting training loop from step {start_step} to {args.num_train_steps}")
+    step = start_step
+    micro_step = 0
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama3_native_te/train_fsdp2_cp.py]: pls add a comment about what `micro_step` is

File: bionemo-recipes/recipes/llama3_native_te/checkpoint.py (hunk 1 above)
Comment (savitha-eng) [reply bionemo-recipes/recipes/llama3_native_te/checkpoint.py:332]: i had the same question in https://github.com/NVIDIA/bionemo-framework/pull/1400 - looks like lots of the changes made there, are reflected here again

**Review by savitha-eng — APPROVED** [review]

File: bionemo-recipes/recipes/llama3_native_te/checkpoint.py (hunk 1 above)
Comment (pstjohn) [reply bionemo-recipes/recipes/llama3_native_te/checkpoint.py:332]: yeah sorry, this one was from #1400. I just needed this for the tests to pass locally, since I think i was getting some race conditions where tests weren't cleaning up correctly

**Comment by pstjohn** [comment]
> Tons of content here

want me to merge #1400 and rebase?

File: bionemo-recipes/recipes/llama3_native_te/train_fsdp2_cp.py (hunk 2 above)
Comment (pstjohn) [reply bionemo-recipes/recipes/llama3_native_te/train_fsdp2_cp.py]: done; this is for gradient accumulation
